### PR TITLE
bug: Allow string quote to be escaped within phpdoc constant

### PR DESCRIPTION
**Pull request by mvorisek**
Reference: phpstan/phpdoc-parser#142 ([1](https://phpstan.org/r/41f1530f-eaee-4e7d-8bfc-e7ccc7cfc7a7), [2](https://phpstan.org/r/a884b37d-682b-45a4-b891-1725ffd1c428)).

### CODE REVIEW

File: src/DocBlock/TypeExpression.php [hunk 1]
```diff
@@ -88,7 +88,7 @@ final class TypeExpression
                     (?i)
```
Comment (keradus) [inline src/DocBlock/TypeExpression.php:88]: I would love to hear more about the purpose of this PR.
looking only at \\\\\\\'"\\'"\'"|"|"\'"\\'\\\\\""\"\""\\\"" is not helping ;)

Comment (mvorisek) [reply src/DocBlock/TypeExpression.php:88]: 1. escaping the quote character was previously not possible
2. `\n` could be possibly reprinted wrongly, thus tested ;-)
3. the test has 2 types, if parsed, they are parsed as two types - this is what the test does

**Comment by keradus** [comment]
@Wirone 
, feeling up for a challenge to review this beauty?

File: tests/DocBlock/TypeExpressionTest.php [hunk 2]
```diff
@@ -146,6 +146,8 @@ public static function provideGetTypesCases(): iterable
         yield ['\\Closure(string|int, bool): bool', ['\\Closure(string|int, bool): bool']];
 
         yield ['array  <  int   , callable  (  string  )  :   bool  >', ['array  <  int   , callable  (  string  )  :   bool  >']];
+
+        yield ['\'a\\\'s"\\\\\n\r\t\'|"b\\"s\'\\\\\n\r\t"', ['\'a\\\'s"\\\\\n\r\t\'', '"b\\"s\'\\\\\n\r\t"']];
```
Comment (kubawerlos) [inline tests/DocBlock/TypeExpressionTest.php:150]: Can we have a few more? What is the smallest accepted type? `'\''`?

**Comment by kubawerlos** [comment]
Why do we need it? If I saw `'` in type in code review I would comment "hey, the type here is broken". Can you add some real-life type test case?

**Comment by Wirone** [comment]
> @Wirone, feeling up for a challenge to review this beauty?

@keradus I agree with @kubawerlos - I don't see the need for this change. Lack of PR description and more examples does not help 😉

**Comment by mvorisek** [comment]
@Wirone please see https://github.com/phpstan/phpdoc-parser/pull/143, escaped phpdoc string is valid and must be supported, it is obvious the original `\'[^\']+?\'` regex cannot handle escaped quotes in quotes.

File: tests/DocBlock/TypeExpressionTest.php (hunk 2 above)
Comment (mvorisek) [reply tests/DocBlock/TypeExpressionTest.php:156]: added smallest type - `''` empty string and a few other simple strin before the complex test

**Comment by Wirone** [comment]
@mvorisek some PRs were merged, could you please resolve the conflicts and update the branch for this PR?

**Comment by mvorisek** [comment]
@Wirone conflict resolved

**Comment by mvorisek** [comment]
@Wirone can this PR be merged?

**Review by kubawerlos — APPROVED** [review]

**Review by Wirone — APPROVED** [review]

**Pull request merged**